src/servers/http_server.cc

```diff
@@ -462,6 +711,250 @@ HTTPAPIServer::Status(h2o_handler_t* handler_self, h2o_req_t* req)
+      v_idx += byte_size;
+    } else {
+      return TRTSERVER_ErrorNew(
+          TRTSERVER_ERROR_INVALID_ARG,
+          std::string(
+              "unexpected size for input '" + io.name() + "', expecting " +
+              std::to_string(byte_size) + " bytes for model '" + model_name +
+              "'")
+              .c_str());
+    }
+  }
+
+  if (v_idx != n) {
+    return TRTSERVER_ErrorNew(
+        TRTSERVER_ERROR_INVALID_ARG,
+        std::string(
+            "unexpected additional input data for model '" + model_name + "'")
+            .c_str());
+  }
+
+  return nullptr;  // success
+}
+
+int
+HTTPAPIServer::Infer(h2o_handler_t* _self, h2o_req_t* req)
+{
+  h2o_custom_req_handler_t* self = (h2o_custom_req_handler_t*)_self;
+  re2::RE2 infer_regex(R"(/api/infer/([^/]+)(?:/(\d+))?)");
+  if (!h2o_memis(req->method.base, req->method.len, H2O_STRLIT("POST"))) {
+    req->res.status = 400;
+    req->res.reason = "Only POST method is allowed";
+    h2o_add_header(
+        &req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL,
+        H2O_STRLIT("text/plain"));
+    h2o_generator_t generator = {NULL, NULL};
+    h2o_iovec_t body = h2o_strdup(&req->pool, "", SIZE_MAX);
+    h2o_start_response(req, &generator);
+    h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
+    return 0;
+  }
+
+  std::string infer_uri =
+      std::string(req->path_normalized.base, req->path_normalized.len);
+  std::string model_name, model_version_str;
+  if ((infer_uri.empty()) ||
+      (!RE2::FullMatch(
+          infer_uri, infer_regex, &model_name, &model_version_str))) {
+    req->res.status = 400;
+    req->res.reason = "Bad Request";
+    h2o_add_header(
+        &req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, NULL,
+        H2O_STRLIT("text/plain"));
+    h2o_generator_t generator = {NULL, NULL};
+    h2o_iovec_t body = h2o_strdup(&req->pool, "", SIZE_MAX);
+    h2o_start_response(req, &generator);
+    h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
+    return 0;
+  }
+
+  int64_t model_version = -1;
+  if (!model_version_str.empty()) {
+    model_version = std::atoll(model_version_str.c_str());
+  }
+
+#ifdef TRTIS_ENABLE_TRACING
+  // Timestamps from evhtp are capture in 'req'. We record here since
+  // this is the first place where we have a tracer.
+  std::unique_ptr<TraceMetaData> trace_meta_data;
+  if (self->http_server->trace_manager_ != nullptr) {
+    trace_meta_data.reset(self->http_server->trace_manager_->SampleTrace());
+    if (trace_meta_data != nullptr) {
+      trace_meta_data->tracer_->SetModel(model_name, model_version);
+      trace_meta_data->tracer_->CaptureTimestamp(
+          TRTSERVER_TRACE_LEVEL_MIN, "http recv start",
+          TIMEVAL_TO_NANOS(req->timestamps.request_begin_at));
+      trace_meta_data->tracer_->CaptureTimestamp(
+          TRTSERVER_TRACE_LEVEL_MIN, "http recv end",
+          TIMEVAL_TO_NANOS(req->timestamps.request_body_begin_at));
```

Comment:
The beginning of the body is not the "http recv end". If there a timestamp or a callback where we can capture the time after the entire http request, including body, has been received?

Reply:
There is no timestamp for capturing the entire request end. There is however a timestamp for when it is done processing the received request. 
https://github.com/h2o/h2o/blob/07459687b819ece4f47ce2c4b5bfd2fae5e287e4/include/h2o.h#L1067
Would this be a more suitable timestamp?

Reply:
Maybe. Or perhaps we should just capture it ourselves when h2o calls use with a complete request.

Reply:
That would be only after the request is _processed_ and the respective handler is 'hit'